numpy/_core/src/multiarray/stringdtype/dtype.c

```diff
@@ -0,0 +1,910 @@
+{
+    static char *kwargs_strs[] = {"", "coerce", "na_object", NULL};
+
+    long size = 0;
+    PyObject *na_object = NULL;
+    int coerce = 1;
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "|l$pO&:StringDType",
+                                     kwargs_strs, &size, &coerce,
+                                     _not_NoValue, &na_object)) {
+        return NULL;
+    }
+
+    return new_stringdtype_instance(na_object, coerce);
+}
+
+static void
+stringdtype_dealloc(PyArray_StringDTypeObject *self)
+{
+    Py_XDECREF(self->na_object);
+    // this can be null if an error happens while initializing an instance
+    if (self->allocator != NULL) {
+        // can we assume the destructor for an instance will only get called
+        // inside of one C thread?
+        NpyString_free_allocator(self->allocator);
+    }
+    PyMem_RawFree((char *)self->na_name.buf);
+    PyMem_RawFree((char *)self->default_string.buf);
+    PyArrayDescr_Type.tp_dealloc((PyObject *)self);
+}
+
+static PyObject *
+stringdtype_repr(PyArray_StringDTypeObject *self)
+{
+    PyObject *ret = NULL;
+    // borrow reference
+    PyObject *na_object = self->na_object;
+    int coerce = self->coerce;
+
+    if (na_object != NULL && coerce == 0) {
+        ret = PyUnicode_FromFormat("StringDType(na_object=%R, coerce=False)",
+                                   na_object);
+    }
+    else if (na_object != NULL) {
+        ret = PyUnicode_FromFormat("StringDType(na_object=%R)", na_object);
+    }
+    else if (coerce == 0) {
+        ret = PyUnicode_FromFormat("StringDType(coerce=False)", coerce);
+    }
+    else {
+        ret = PyUnicode_FromString("StringDType()");
+    }
+
+    return ret;
+}
+
+static PyObject *_convert_to_stringdtype_kwargs = NULL;
+
+static PyObject *
+stringdtype__reduce__(PyArray_StringDTypeObject *self, PyObject *NPY_UNUSED(args))
+{
+    npy_cache_import("numpy._core._internal", "_convert_to_stringdtype_kwargs",
+                     &_convert_to_stringdtype_kwargs);
+
+    if (_convert_to_stringdtype_kwargs == NULL) {
+        return NULL;
+    }
+
+    if (self->na_object != NULL) {
+        return Py_BuildValue("O(iO)", _convert_to_stringdtype_kwargs,
+                             self->coerce, self->na_object);
+    }
+
+    return Py_BuildValue("O(i)", _convert_to_stringdtype_kwargs, self->coerce);
+}
+
+static PyMethodDef PyArray_StringDType_methods[] = {
+        {
+                "__reduce__",
+                (PyCFunction)stringdtype__reduce__,
+                METH_NOARGS,
+                "Reduction method for a StringDType object",
+        },
+        {NULL, NULL, 0, NULL},
+};
+
+static PyMemberDef PyArray_StringDType_members[] = {
+        {"na_object", T_OBJECT_EX, offsetof(PyArray_StringDTypeObject, na_object),
+         READONLY,
+         "The missing value object associated with the dtype instance"},
+        {"coerce", T_INT, offsetof(PyArray_StringDTypeObject, coerce), READONLY,
```

Comment:
Good call! I made all the other flag members of the `PyArray_StringDTypeObject` struct `char`s too while I was at it and snuck in a NEP update too because the NEP includes a verbatim copy of the struct layout.